File: core/src/com/tiles/pathfinding/Pawn.java
package com.tiles.pathfinding;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.math.Vector2;

import java.io.Serializable;
import java.util.ArrayList;

import static com.tiles.pathfinding.NeededConstants.mouseInput;
import static com.tiles.pathfinding.NeededConstants.tileList;

public class Pawn implements Serializable {
    private String color; // La couleur du pion

    public String getColor() {
        return color;
    }

    public Case setCase; // La case sur laquelle est le pion
    private transient Sprite sprite;

    Pawn(String color) {
        this.color = color;
    }

    private float size;
    // Je le met parce que Compte me l'a demandé
    // mais je pense pas que ça vaille le coup vu qu'il est forcèment chargé ?

    private void setSize(float size) {
        this.size = size;
        sprite.setSize(size, 2 * size);
    }

    public void setSize() {
        setSize(setCase.caseSize() / 2);
    }
    // Fonctions classiques pour gérer la taille

    private boolean isMovable = false; // Même principe que pour la Queue

    public void load() { // Pour la sérialization
        sprite = new Sprite(new Texture("pions/" + color + ".png"));
        setSize();
        updateCoordinates();
    }

    public void draw() {
        sprite.draw(NeededConstants.batch);
    }

    public void updateCoordinates() {
        sprite.setX(setCase.getX(setCase.getRotatedCoordinates()[0]) + (setCase.caseSize() - sprite.getWidth()) / 2);
        sprite.setY(setCase.getY(setCase.getRotatedCoordinates()[1]) + setCase.caseSize() / 3);
    }

    public void dispose() {
        sprite.getTexture().dispose();
    }

    private boolean hasExplored = false; // Pour regarder si on a déjà fait le pathfinding et éviter de le faire trop de fois

    public void handleInput(Player player) {
        if (isMovable) {
            sprite.setX(mouseInput().x - sprite.getWidth() / 2);
            sprite.setY(mouseInput().y - sprite.getHeight() / 2);
            if (!hasExplored) { // On fait le pathfinding une seule fois
                setCase.show(); // On montre la case de départ
                setCase.explore(player); // On lance le pathfinding (structure récursive)
                hasExplored = true; // Et on montre qu'on a déjà fait le pathfinding
            }
            if (Gdx.input.isButtonJustPressed(Input.Buttons.LEFT)) { // Puis si on clique quelque part
                try {
                    Case nextCase = findCase(); // Est-on sur une case ?
                    if (nextCase.isValid) { // Si elle existe et est non nulle
                        setCase.revert(player); // On annule le pathfinding... avec un autre pathfinding
                        setCase.hide(); // On cache la case de départ
                        setCase = nextCase; // On change la case
                        hasExplored = false; // On réinitialise les variables booléennes
                        isMovable = false;
                        updateCoordinates(); // Et on met à jour les coordonées, qui sont entièrement calculées à partir de la case
                    }
                } catch (NullPointerException e) {
                    System.out.println("No Tile found in Pawn.handleInput");
                }
            }
        } else isMovable = (Gdx.input.isButtonJustPressed(Input.Buttons.RIGHT) &&
                (sprite.getX() < mouseInput().x) && (mouseInput().x < sprite.getX() + sprite.getWidth() &&
                (sprite.getY() < mouseInput().y) && (mouseInput().y < sprite.getY() + sprite.getHeight())));
    }

    public Case findCase() {
        for (Tile tile : tileList) {
            if (tile.getX() <= mouseInput().x && mouseInput().x <= tile.getX() + tile.getSize() &&
                    tile.getY() <= mouseInput().y && mouseInput().y <= tile.getY() + tile.getSize()) {
                try {
                    return tile.getCase(mouseInput());
                } catch (ArrayIndexOutOfBoundsException e) {
                    System.out.println("Clicked border of Tile n°" + tile.number + " in Pawn.findCase");
                }
            }
        }
        return null;
    }
}

File: core/src/com/mygdx/game/Pawn.kt
package com.mygdx.game


import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Vector3
import java.lang.System.load

class Pawn (Texture: Texture,
           var x: Float, var y: Float,
           var width: Float, var height: Float, val pioncolor: String) {

    val colonne0: IntArray = intArrayOf(0,1,1,0,1)
    val colonne1: IntArray = intArrayOf(0,1,1,0,1)
    val colonne2: IntArray = intArrayOf(0,1,1,0,1)
    val colonne3: IntArray = intArrayOf(0,0,0,0,1)
    val colonne4: IntArray = intArrayOf(1,1,1,1,1)
    public val tab: Array<IntArray> = arrayOf(colonne0,colonne1,colonne2,colonne3,colonne4)

    var isMovable = false

    var highlight: Sprite = Sprite(Texture("highlight.png"))

    val texture: Sprite = Sprite(Texture)

    init {
        updateSprite()
    }

    fun updateSprite() {
        texture.setPosition(x, y)
        highlight.setPosition(x-25,y-25)
        texture.setSize(width, height)

    }
    //donne la position au pion
    fun setPosition(x: Float, y: Float) {
        this.x = x
        this.y = y
    }
    //donne la taille du sprite
    fun setSize(width: Float, height: Float) {
        this.width = width
        this.height = height
    }

    fun update(batch: SpriteBatch) {
        texture.draw(batch)
        highlight.draw(batch)
    }

    //dit si le pion est cliqué
    //fun isClicked(iX: Float, iY: Float): Boolean = (x <= iX && iX <= x + width) && (y <= iY && iY <= y + height)

    fun movePawn(inputX: Float, inputY: Float) {
        val highlight = Gdx.files.internal("highlight.png")
        var nbX = (inputX/100).toInt()
        var nbY = (inputY/100).toInt()
        var X=(nbX*100+25).toFloat()
        var Y=(nbY*100+25).toFloat()
        if (isMovable) {
            if (tab[nbX][nbY] == 0) {
                setPosition(X, Y)
                updateSprite()
                isMovable = false
                if (Gdx.input.isButtonJustPressed(Input.Buttons.LEFT)) {
                    isMovable = false
                }
            }
        } else {
            isMovable = isMovable || (Gdx.input.isButtonJustPressed(Input.Buttons.LEFT) && (x < inputX) && (inputX < x + width && (y < inputY) && (inputY < y + height)))

        }
    }

    fun dispose() {
        texture.texture.dispose()
        //highlight.texture.dispose()
    }

    fun getColor(): String = pioncolor
}













